Cabal/src/Distribution/PackageDescription/Check.hs

```diff
@@ -1966,25 +1972,41 @@ repoTypeDirname Pijul     = [".pijul"]
 -- should be done for example when creating or validating a package tarball.
 --
 checkPackageFileNames :: [FilePath] -> [PackageCheck]
-checkPackageFileNames files =
-     (take 1 . mapMaybe checkWindowsPath $ files)
-  ++ (take 1 . mapMaybe checkTarPath     $ files)
-      -- If we get any of these checks triggering then we're likely to get
-      -- many, and that's probably not helpful, so return at most one.
-
-checkWindowsPath :: FilePath -> Maybe PackageCheck
-checkWindowsPath path =
-  check (not $ FilePath.Windows.isValid path') $
-    PackageDistInexcusable $
-         "Unfortunately, the file " ++ quote path ++ " is not a valid file "
-      ++ "name on Windows which would cause portability problems for this "
-      ++ "package. Windows file names cannot contain any of the characters "
-      ++ "\":*?<>|\" and there are a few reserved names including \"aux\", "
-      ++ "\"nul\", \"con\", \"prn\", \"com1-9\", \"lpt1-9\" and \"clock$\"."
+checkPackageFileNames = checkPackageFileNamesWithGlob . zip (repeat True)
+
+checkPackageFileNamesWithGlob :: [(Bool, FilePath)] -> [PackageCheck]
+checkPackageFileNamesWithGlob files =
+  catMaybes $
+    checkWindowsPaths files
+    :
+    [ checkTarPath file
+    | (_, file) <- files
+    ]
+
+checkWindowsPaths :: [(Bool, FilePath)] -> Maybe PackageCheck
+checkWindowsPaths paths =
+    case filter (not . FilePath.Windows.isValid . escape) paths of
+      [] -> Nothing
+      ps -> Just $
+        PackageDistInexcusable $
+             "Unfortunately, the " ++ quotes (map snd ps) ++ " on Windows "
+          ++ "which would cause portability problems for this package. "
```

Comment:
Not sure how to improve the wording. Would "paths" instead of "the files" be as accurate? Is 'C:/foo" a path or is it a drive and a path? Also, I guess there should be a comma before 'which'?

Reply:
How does "the path(s) ... is/are invalid on Windows, which" sound?  Getting into whether the drive is separate from the path is probably a bit more detail than necessary (most users would just consider it all a single path, correct or not), especially since the vast majority of paths in .cabal files are required to be relative.  We should also probably remove the "Unfortunately" to better fit the other warning messages; the *only* other ones which use it describe needing to specify a newer `cabal-version`.  The below is also probably worth comparing and updating for more consistency (if nothing else, not everything that fails is "reserved" per se):

https://github.com/haskell/cabal/blob/bb516ed31bf31758627e25a663071ac4d2acc8ba/Cabal/src/Distribution/PackageDescription/Check.hs#L441-L446

Reply:
Agreed on all accounts.